THRIFT-5093: lib: cpp: test: clarify effect of MemoryPolicy on TMemoryBuffer

cfriedt opened the pull request:
<!-- Explain the changes in the pull request below: -->

My main use case with the original code change was that I wanted to be able to call `resetBuffer()` on `OBSERVE` buffers and reuse the contents for test cases over and over. While the default `resetBuffer()` cannot be used for that purpose, the same result can be achieved by calling `resetBuffer(ro_mem, ro_mem_size, OBSERVE)` as well, so the workaround is fairly trivial.

While the former implementation would be more convenient in some cases, it makes more sense to have consistent behaviour for all `resetBuffer()` implementations.

Added test cases to clarify the effect of `MemoryPolicy` on  `TMemoryBuffer` behaviour.

<!-- We recommend you review the checklist/tips before submitting a pull request. -->

- [x] Did you create an [Apache Jira](https://issues.apache.org/jira/projects/THRIFT/issues/) ticket?  (not required for trivial changes)
- [x] If a ticket exists: Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?
- [x] Did you squash your changes to a single commit?  (not required, but preferred)
- [x] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"?
- [ ] If your change does not involve any code, add ` [skip ci]` at the end of your pull request to free up build resources.

<!--
  The Contributing Guide at:
  https://github.com/apache/thrift/blob/master/CONTRIBUTING.md
  has more details and tips for committing properly.
-->

cfriedt commented:
Not sure why multiple builds are failing, but it's highly unlikely it has anything to do with this change.

Jens-G commented:
There's also #2030 ... may I ask to cross-review each others change, resolving any potential interferences?

emmenlau commented:
Dear @cfriedt I've just reviewed your change and I'm not certain if it is correct. Could you motivate your change further?

In my understanding, *all* init methods in the constructor `TMemoryBuffer(uint8_t* buf, uint32_t sz, MemoryPolicy policy = OBSERVE)` from existing memory will move the internal `rBound_` and `wBound_` up by `sz`. In the case of `OBSERVE` or `TAKE_OWNERSHIP` this happens in `initCommon()` whereas in the case of `COPY` this (probably) happens in `write()`.
Your PR changes this behavior. I can not say with certainty if or why the current behavior is correct, but I' under the impression that it was written consistently with a certain intention in mind. And your PR breaks this consistency.

Or am I on the wrong track? Feedback welcome!

cfriedt commented:
> Or am I on the wrong track? Feedback welcome!

I think there is some conflation about whether `wPos` (and by extension `wBase_` and `wBound_` has any bearing on whether or not the buffer is writeable.

`wPos` does not determine if the buffer is writeable (in terms of permissions). It's `owner_` that is responsible for that, as evident in `TMemoryBuffer::ensureCanWrite()`.

emmenlau commented:
> I think there is some conflation about whether `wPos` (and by extension `wBase_` and `wBound_`) has any bearing on whether or not the buffer is writeable.

I can not speak for other people and I don't know what has been discussed so far. However my concern is not about buffer ownership. I quote the relevant section:
```cpp
    rBound_ = buffer_ + wPos;
    // TODO(dreiss): Investigate NULL-ing this if !owner.
    wBase_ = buffer_ + wPos;
```
So `wPos` is used to move `rBound_` and `wBase_` forward to the end of the buffer in all cases where `TMemoryBuffer` is constructed from existing memory. This makes certain sense to me, because its likely that users do not want to overwrite existing data.

I'm more than happy to get educated on the meaning of `rBound_` and `wBase_` and why you think the current implementation is incorrect. But at least from my limited understanding I am not convinced that this change is correct.

cfriedt commented:
> I can not speak for other people and I don't know what has been discussed so far. However my concern is not about buffer ownership. I quote the relevant section:
> 
> ```c++
>     rBound_ = buffer_ + wPos;
>     // TODO(dreiss): Investigate NULL-ing this if !owner.
>     wBase_ = buffer_ + wPos;
> ```
> 
> So `wPos` is used to move `rBound_` and `wBase_` forward to the end of the buffer in all cases where `TMemoryBuffer` is constructed from existing memory. This makes certain sense to me, because its likely that users do not want to overwrite existing data.

What if there is no existing data to overwrite? Then the buffer *always* has zero space. available for writing.

I think I'm beginning to see your point though; by fixing one use case, it breaks another.

Just to clarify, the use case that I have which is currently broken is this:
* a pre-allocated, fixed-size buffer is passed in
* The TMemoryBuffer should take ownership of the pre-allocated buffer
* no data has been written to the buffer
* all subsequent writes to the TMemoryBuffer fail, because `wBase_ == wBound_`

The use case that you are saying would be broken by this change is:
* a pre-allocated, fixed-size buffer is passed in
* The TMemoryBuffer should only observe the pre-allocated buffer
* irrelevant whether data has been written to the buffer or not
* all subsequent writes to the TMemoryBuffer fail, because `wBase_ == wBound_`


> I'm more than happy to get educated on the meaning of `rBound_` and `wBase_` and why you think the current implementation is incorrect. But at least from my limited understanding I am not convinced that this change is correct.

Is it possible that the correct line should be the following? From what I can tell, it fixes both uses cases listed above.
```
initCommon(buf, sz, policy == TAKE_OWNERSHIP,  ( policy == TAKE_OWNERSHIP ) ? 0 : sz );
```

cfriedt commented:
pushing back again, sorry

Jens-G commented:
Not sure about the state of this, also CI seems all "red". How do we proceed here?

cfriedt commented:
Let me take a look again.

cfriedt commented:
@Jens-G - should be good now. Would you like to assign reviewers?

cfriedt commented:
AppVeyor is failing for completely unrelated reasons ("cannot used reserved language keyword return").

Not even sure where to begin with Travis.

Re-running tests locally on aarch64 and amd64.

cfriedt commented:
```console
$ make check
...
make[5]: Leaving directory '/home/cfriedt/workspace/thrift/lib/cpp/test'
make  check-TESTS
make[5]: Entering directory '/home/cfriedt/workspace/thrift/lib/cpp/test'
An error message from getaddrinfo on the console is expected:
Thrift: Fri Jun 19 07:12:34 2020 getaddrinfo() -> -2; Name or service not known

*** No errors detected
PASS: UnitTests
Thrift: Fri Jun 19 07:12:34 2020 ~TFDTransport TTransportException: 'TFDTransport::close(): Bad file descriptor'

*** No errors detected
PASS: TFDTransportTest

*** No errors detected
PASS: TPipedTransportTest
*** No errors detected
PASS: DebugProtoTest

*** No errors detected
PASS: JSONProtoTest

*** No errors detected
PASS: OptionalRequiredTest

*** No errors detected
PASS: RecursiveTest

*** No errors detected
PASS: SpecializationTest
TBinaryProtocol => OK
TLEBinaryProtocol => OK
TCompactProtocol => OK

*** No errors detected
PASS: AllProtocolsTest
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport

*** No errors detected
PASS: TransportTest
Thrift: Fri Jun 19 07:13:21 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)

*** No errors detected
PASS: TInterruptTest

*** No errors detected
PASS: TServerIntegrationTest
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)

*** No errors detected
PASS: SecurityTest
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Fri Jun 19 07:13:40 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)

*** No errors detected
PASS: SecurityFromBufferTest
seed: 1592565220

*** No errors detected
PASS: ZlibTest

*** No errors detected
PASS: TFileTransportTest
PASS: link_test

*** No errors detected
PASS: OpenSSLManualInitTest

*** No errors detected
PASS: EnumTest

*** No errors detected
PASS: RenderedDoubleConstantsTest

*** No errors detected
PASS: AnnotationTest
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: join done for IO thread #0
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: join done for IO thread #0
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: join done for IO thread #0

*** No errors detected
PASS: TNonblockingServerTest
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Fri Jun 19 07:13:43 2020 TConnection::workSocket(): client disconnected
Thrift: Fri Jun 19 07:13:43 2020 SSL_shutdown: Broken pipe (SSL_error_code = 6)
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: join done for IO thread #0
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Fri Jun 19 07:13:43 2020 TConnection::workSocket(): client disconnected
Thrift: Fri Jun 19 07:13:43 2020 SSL_shutdown: Broken pipe (SSL_error_code = 6)
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: join done for IO thread #0
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Fri Jun 19 07:13:43 2020 TConnection::workSocket(): client disconnected
Thrift: Fri Jun 19 07:13:43 2020 SSL_shutdown: Broken pipe (SSL_error_code = 6)
Thrift: Fri Jun 19 07:13:43 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Fri Jun 19 07:13:43 2020 TNonblocking: join done for IO thread #0

*** No errors detected
PASS: TNonblockingSSLServerTest
===================
All 23 tests passed
===================
...
cfriedt@cfriedt-MacBookPro:~/workspace/thrift$ echo $?
0
cfriedt@cfriedt-MacBookPro:~/workspace/thrift$ uname -a
Linux cfriedt-MacBookPro 4.15.0-106-generic #107-Ubuntu SMP Thu Jun 4 11:27:52 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

cfriedt commented:
On aarch64, completely unrelated failures (also works on x86_64, but not on aarch64). Error is specific to SSL.

x86_64 machine is Ubuntu Bionic, aarch64 machine is Debian 10.2

x86_64: libssl1.1 is already the newest version (1.1.1-1ubuntu2.1~18.04.6).
aarch64: libssl1.1 is already the newest version (1.1.1d-0+deb10u3).

```console
make[5]: Leaving directory '/home/chrisfriedt/workspace/thrift/lib/cpp/test'
make  check-TESTS
make[5]: Entering directory '/home/chrisfriedt/workspace/thrift/lib/cpp/test'
Thrift: Tue Jun 16 07:16:22 2020 TSocket::open() connect() <Host: localhost Port: 46709>: Connection refused
An error message from getaddrinfo on the console is expected:
Thrift: Tue Jun 16 07:16:26 2020 getaddrinfo() -> -2; Name or service not known

*** No errors detected
PASS: UnitTests
Thrift: Tue Jun 16 07:16:26 2020 ~TFDTransport TTransportException: 'TFDTransport::close(): Bad file descriptor'

*** No errors detected
PASS: TFDTransportTest

*** No errors detected
PASS: TPipedTransportTest

*** No errors detected
PASS: DebugProtoTest

*** No errors detected
PASS: JSONProtoTest

*** No errors detected
PASS: OptionalRequiredTest

*** No errors detected
PASS: RecursiveTest

*** No errors detected
PASS: SpecializationTest
TBinaryProtocol => OK
TLEBinaryProtocol => OK
TCompactProtocol => OK

*** No errors detected
PASS: AllProtocolsTest
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport
Timeout alarm expired; attempting to unblock transport

*** No errors detected
PASS: TransportTest
Thrift: Tue Jun 16 07:17:29 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)

*** No errors detected
PASS: TInterruptTest

*** No errors detected
PASS: TServerIntegrationTest
Thrift: Tue Jun 16 07:17:49 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Tue Jun 16 07:17:49 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Tue Jun 16 07:17:49 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Tue Jun 16 07:17:49 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
terminate called after throwing an instance of 'boost::execution_aborted'

*** 4 failures are detected in the test module "SecurityTest"
FAIL: SecurityTest
Thrift: Tue Jun 16 07:17:49 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Tue Jun 16 07:17:49 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Tue Jun 16 07:17:49 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Tue Jun 16 07:17:49 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
terminate called after throwing an instance of 'boost::execution_aborted'

*** 4 failures are detected in the test module "SecurityFromBufferTest"
FAIL: SecurityFromBufferTest
seed: 1592306269

*** No errors detected
PASS: ZlibTest

*** No errors detected
PASS: TFileTransportTest
PASS: link_test

*** No errors detected
PASS: OpenSSLManualInitTest
*** No errors detected
PASS: EnumTest

*** No errors detected
PASS: RenderedDoubleConstantsTest

*** No errors detected
PASS: AnnotationTest
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Tue Jun 16 07:17:56 2020 TSocket::open() connect() <Host: localhost Port: 12345>: Connection refused
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: join done for IO thread #0
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Tue Jun 16 07:17:56 2020 TSocket::open() connect() <Host: localhost Port: 33643>: Connection refused
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: join done for IO thread #0
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Tue Jun 16 07:17:56 2020 TSocket::open() connect() <Host: localhost Port: 37905>: Connection refused
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: join done for IO thread #0

*** No errors detected
PASS: TNonblockingServerTest
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Tue Jun 16 07:17:56 2020 TSocket::open() connect() <Host: localhost Port: 12345>: Connection refused
Thrift: Tue Jun 16 07:17:56 2020 TConnection::workSocket(): client disconnected
Thrift: Tue Jun 16 07:17:56 2020 SSL_shutdown: Broken pipe (SSL_error_code = 6)
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: join done for IO thread #0
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Tue Jun 16 07:17:56 2020 TSocket::open() connect() <Host: localhost Port: 36637>: Connection refused
Thrift: Tue Jun 16 07:17:56 2020 TConnection::workSocket(): client disconnected
Thrift: Tue Jun 16 07:17:56 2020 SSL_shutdown: Broken pipe (SSL_error_code = 6)
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: join done for IO thread #0
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: Serving with 1 io threads.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: using libevent 2.1.8-stable method epoll
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for listen.
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: IO thread #0 registered for notify.
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 entering loop...
Thrift: Tue Jun 16 07:17:56 2020 TSocket::open() connect() <Host: localhost Port: 40159>: Connection refused
Thrift: Tue Jun 16 07:17:56 2020 TConnection::workSocket(): client disconnected
Thrift: Tue Jun 16 07:17:56 2020 SSL_shutdown: Broken pipe (SSL_error_code = 6)
Thrift: Tue Jun 16 07:17:56 2020 TNonblockingServer: IO thread #0 run() done!
Thrift: Tue Jun 16 07:17:56 2020 TNonblocking: join done for IO thread #0

*** No errors detected
PASS: TNonblockingSSLServerTest
====================
2 of 23 tests failed
====================

 ./SecurityTest
Running 1 test case...
Thrift: Tue Jun 16 07:23:33 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Tue Jun 16 07:23:33 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
SecurityTest.cpp(268): error: in "SecurityTest/ssl_security_matrix":       Server = SSLTLS, Client = TLSv1_0 expected mConnected == 1 but was 0
Thrift: Tue Jun 16 07:23:33 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Tue Jun 16 07:23:33 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
SecurityTest.cpp(268): error: in "SecurityTest/ssl_security_matrix":       Server = SSLTLS, Client = TLSv1_1 expected mConnected == 1 but was 0
SecurityTest.cpp(149): fatal error: in "SecurityTest/ssl_security_matrix": N6apache6thrift9transport13TSSLExceptionE: SSL_CTX_set_cipher_list: bad value
terminate called after throwing an instance of 'boost::execution_aborted'
unknown location(0): fatal error: in "SecurityTest/ssl_security_matrix": signal: SIGABRT (application abort requested)
SecurityTest.cpp(149): last checkpoint

*** 4 failures are detected in the test module "SecurityTest"

$ ./SecurityFromBufferTest
Running 1 test case...
Thrift: Tue Jun 16 07:19:42 2020 SSL_shutdown: Broken pipe (SSL_error_code = 5)
Thrift: Tue Jun 16 07:19:42 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Tue Jun 16 07:19:42 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
SecurityFromBufferTest.cpp(245): error: in "SecurityFromBufferTest/ssl_security_matrix":       Server = SSLTLS, Client = TLSv1_0 expected mConnected == 1 but was 0
Thrift: Tue Jun 16 07:19:42 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
Thrift: Tue Jun 16 07:19:42 2020 SSL_shutdown: shutdown while in init (SSL_error_code = 1)
SecurityFromBufferTest.cpp(245): error: in "SecurityFromBufferTest/ssl_security_matrix":       Server = SSLTLS, Client = TLSv1_1 expected mConnected == 1 but was 0
SecurityFromBufferTest.cpp(145): fatal error: in "SecurityFromBufferTest/ssl_security_matrix": N6apache6thrift9transport13TSSLExceptionE: SSL_CTX_set_cipher_list: bad value
terminate called after throwing an instance of 'boost::execution_aborted'
unknown location(0): fatal error: in "SecurityFromBufferTest/ssl_security_matrix": signal: SIGABRT (application abort requested)
SecurityFromBufferTest.cpp(145): last checkpoint
$ uname -a
Linux penguin 4.19.113-08528-g5803a1c7e9f9 #1 SMP PREEMPT Thu Apr 2 15:16:47 PDT 2020 aarch64 GNU/Linux
```

cfriedt commented:
Added `test_take_ownership()` to illustrate that is still working fine.

cfriedt commented:
Reviewers, @emmenlau  @Jens-G?

cfriedt commented:
Also, this should probably be labelled as a bug and not a feature (it has no labels yet).

Jens-G commented:
@emmenlau Can do do that?

emmenlau commented on lib/cpp/src/thrift/transport/TBufferTransports.h:
```diff
@@ -587,11 +587,9 @@ class TMemoryBuffer : public TVirtualTransport<TMemoryBuffer, TBufferBase> {
-    if (!owner_) {
-      wBound_ = wBase_;
-      bufferSize_ = 0;
+    if (owner_) {
+      wBase_ = buffer_;
+      wBound_ = buffer_ + bufferSize_;
```
Could you explain this change a bit? Originally the code always set `wBase_ = buffer_;` but after your change, `wBase_` is only set if the buffer is owned. Is that guaranteed correct?

cfriedt replied:
If the buffer is not owned (read-only), wbase is initialized to buffer + wpos (the end of the buffer) and never updated, along with wbound. Throughout the lifetime of the buffer, if it is not owned (read-only), wbase == wbound meaning there is no space to write. Only if the buffer is owned (writeable) should wbase and wbound be reset to the beginning / end of the buffer.

cfriedt replied:
Please let me know if you would like additional testing done to illustrate things a bit more.

Jens-G commented:
@emmenlau Still not sure what the status here is ... :-)

Jens-G commented:
@emmenlau Ping?

cfriedt commented:
I could try rebasing to see if unrelated tests pass.

cfriedt commented:
> I could try rebasing to see if unrelated tests pass.

It's already at the latest from master.

cfriedt commented:
@Jens-G - will this be merged soon? I'd be happy to make additional changes if required.

emmenlau commented:
I'll try to review this again soon. Please ping me if I don't get back to it soon!

cfriedt commented:
Ping, @emmenlau ;-)

emmenlau commented:
@cfriedt I'll pull this into my local build and see how it performs, ok? Its not a merge yet, but at least a step on the way. Thanks again!

emmenlau commented:
I've been using these changes successfully in my code for a few months now. I'm not certain that qualifies as a validation, but I personally would be open to merging this PR. Other opinions?

emmenlau commented:
@cfriedt for the sake of a nicer git history, could you rebase on latest master?
@all, I would merge these changes as they have worked for me successfully for a long time and there are tests that seem to confirm the improved behavior. Please speak up if there are any concerns against merging.

cfriedt commented:
@emmenlau - will do

emmenlau commented:
@cfriedt I've just this very second merged a PR that fixes the Travis build, please be sure to include it in your rebase if possible. Thanks a lot!

emmenlau commented:
Ok @cfried, I'm sorry but there is indeed a problem: With your changes, there are consistent crashes on all Windows builds (MSVC and MINGW) in `processor_test`, see https://ci.appveyor.com/project/ApacheSoftwareFoundation/thrift/builds/40571399. With another very recent build from yesterday, we do not get these, see https://ci.appveyor.com/project/ApacheSoftwareFoundation/thrift/builds/40603625.

You can try to rebase again on latest master and force push, just to be sure its not a random fluctuation. But if the Windows tests fail again, we will need to follow it up (or at least not merge).

cfriedt commented:
> Ok @cfried, I'm sorry but there is indeed a problem: With your changes, there are consistent crashes on all Windows builds (MSVC and MINGW) in `processor_test`, see https://ci.appveyor.com/project/ApacheSoftwareFoundation/thrift/builds/40571399. With another very recent build from yesterday, we do not get these, see https://ci.appveyor.com/project/ApacheSoftwareFoundation/thrift/builds/40603625.

Odd - yeah, the logs show this:
```
[00:15:37] 397: Test command: C:\projects\build\MSVC2017\x64\bin\Release\processor_test.exe
[00:15:37] 397: Test timeout computed to be: 300
[00:15:39] 397: Thrift: Mon Aug 30 10:33:14 2021 TNonblockingServer: Serving with 1 io threads.
[00:15:39] 397: Thrift: Mon Aug 30 10:33:14 2021 TNonblockingServer: using libevent 2.1.8-stable method win32
[00:15:39] 397: Thrift: Mon Aug 30 10:33:14 2021 TNonblocking: IO thread #0 registered for listen.
[00:15:39] 397: Thrift: Mon Aug 30 10:33:14 2021 TNonblocking: IO thread #0 registered for notify.
[00:15:39] 397: Thrift: Mon Aug 30 10:33:14 2021 TNonblockingServer: IO thread #0 entering loop...
[00:17:43] 397: Thrift: Mon Aug 30 10:35:18 2021 TNonblockingServer: caught bad_alloc exception.
[00:17:44] 397: Running 64 test cases...
[00:17:46] 397/416 Test #397: processor_test ...................***Failed  128.95 sec
```

I'm not sure where the bad_alloc exception would come from, but I'll look into it if it does not pass after rebasing.

emmenlau commented:
Hi, any updates on this?

cfriedt commented:
Man - I wish there were more hours in the day! Up until a few weeks ago, I only had Mac machines and it was somewhat painful to set up Windows in a VM to try and reproduce the weird Windows issues. Now I have a decent Linux machine to use for that, so I'll give it another go in the next few days.

Thanks, for checking in :-)

emmenlau commented:
> Man - I wish there were more hours in the day!

Haha, don't we all! Happy coding and keep up the good work! Would be great if this comes to a good solution...

cfriedt commented:
@emmenlau - repurposing this ticket to simply add clarifying tests.

My main use case with the original code change was that I wanted to be able to call `resetBuffer()` on `OBSERVE` buffers and reuse them for test cases. I can achieve that by calling `resetBuffer(ro_mem, ro_mem_size, OBSERVE)` as well, so the workaround is fairly trivial. It makes more sense to have consistent behaviour for all `resetBuffer()` implementations.

Jens-G commented:
There are a few errors and warnings, could you have a look at that?
https://ci.appveyor.com/project/ApacheSoftwareFoundation/thrift/builds/43268999/job/2fkl3a5gaf0mlduy

cfriedt commented:
> There are a few errors and warnings, could you have a look at that? https://ci.appveyor.com/project/ApacheSoftwareFoundation/thrift/builds/43268999/job/2fkl3a5gaf0mlduy

Done! As it turns out, MSVC `constexpr` support might not be as complete as in g++ or clang++.

https://docs.microsoft.com/en-us/cpp/cpp/lambda-expressions-constexpr?view=msvc-170

cfriedt commented:
Thanks for your patience & reviews @Jens-G & @emmenlau

emmenlau commented:
Thanks to you for bringing this home!
